File: sdk-blocking/src/main/java/dev/restate/sdk/blocking/RestateContextImpl.java
package dev.restate.sdk.blocking;

import com.google.protobuf.MessageLite;
import dev.restate.generated.core.AwakeableIdentifier;
import dev.restate.sdk.core.StateKey;
import dev.restate.sdk.core.TypeTag;
import dev.restate.sdk.core.syscalls.*;
import io.grpc.MethodDescriptor;
import io.grpc.StatusRuntimeException;
import java.time.Duration;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.function.Supplier;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

class RestateContextImpl implements RestateContext {

  private final Syscalls syscalls;

  RestateContextImpl(Syscalls syscalls) {
    this.syscalls = syscalls;
  }

  @Override
  public <T> Optional<T> get(StateKey<T> key) {
    DeferredResult<T> deferredResult =
        Util.blockOnSyscall(cb -> syscalls.get(key.name(), key.typeTag(), cb));

    if (!deferredResult.isCompleted()) {
      Util.<Void>blockOnSyscall(cb -> syscalls.resolveDeferred(deferredResult, cb));
    }

    return Util.unwrapOptionalReadyResult(deferredResult.toReadyResult());
  }

  @Override
  public void clear(StateKey<?> key) {
    Util.<Void>blockOnSyscall(cb -> syscalls.clear(key.name(), cb));
  }

  @Override
  public <T> void set(StateKey<T> key, @Nonnull T value) {
    Util.<Void>blockOnSyscall(cb -> syscalls.set(key.name(), key.typeTag(), value, cb));
  }

  @Override
  public Awaitable<Void> timer(Duration duration) {
    DeferredResult<Void> result = Util.blockOnSyscall(cb -> syscalls.sleep(duration, cb));
    return new Awaitable<>(syscalls, result);
  }

  @Override
  public <T extends MessageLite, R extends MessageLite> Awaitable<R> call(
      MethodDescriptor<T, R> methodDescriptor, T parameter) {
    DeferredResult<R> result =
        Util.blockOnSyscall(cb -> syscalls.call(methodDescriptor, parameter, cb));
    return new Awaitable<>(syscalls, result);
  }

  @Override
  public <T extends MessageLite> void backgroundCall(
      MethodDescriptor<T, ? extends MessageLite> methodDescriptor, T parameter) {
    Util.<Void>blockOnSyscall(cb -> syscalls.backgroundCall(methodDescriptor, parameter, cb));
  }

  @Override
  public <T> T sideEffect(TypeTag<T> typeTag, Supplier<T> action) {
    CompletableFuture<CompletableFuture<T>> enterFut = new CompletableFuture<>();
    syscalls.enterSideEffectBlock(
        typeTag,
        new EnterSideEffectSyscallCallback<>() {
          @Override
          public void onNotExecuted() {
            enterFut.complete(new CompletableFuture<>());
          }

          @Override
          public void onResult(T result) {
            enterFut.complete(CompletableFuture.completedFuture(result));
          }

          @Override
          public void onFailure(StatusRuntimeException t) {
            enterFut.complete(CompletableFuture.failedFuture(t));
          }

          @Override
          public void onCancel(Throwable t) {
            enterFut.cancel(true);
          }
        });

    // If a failure was stored, it's simply thrown here
    CompletableFuture<T> exitFut = Util.awaitCompletableFuture(enterFut);
    if (exitFut.isDone()) {
      // We already have a result, we don't need to execute the action
      return Util.awaitCompletableFuture(exitFut);
    }

    ExitSideEffectSyscallCallback<T> exitCallback =
        new ExitSideEffectSyscallCallback<>() {
          @Override
          public void onResult(T result) {
            exitFut.complete(result);
          }

          @Override
          public void onFailure(StatusRuntimeException t) {
            exitFut.completeExceptionally(t);
          }

          @Override
          public void onCancel(@Nullable Throwable t) {
            exitFut.cancel(true);
          }
        };

    T res = null;
    Throwable failure = null;
    try {
      res = action.get();
    } catch (Throwable e) {
      failure = e;
    }

    if (failure != null) {
      syscalls.exitSideEffectBlockWithException(failure, exitCallback);
    } else {
      syscalls.exitSideEffectBlock(typeTag, res, exitCallback);
    }

    return Util.awaitCompletableFuture(exitFut);
  }

  @Override
  public <T> Awakeable<T> awakeable(TypeTag<T> typeTag) throws StatusRuntimeException {
    // Retrieve the awakeable
    Map.Entry<AwakeableIdentifier, DeferredResult<T>> awakeable =
        Util.blockOnSyscall(cb -> syscalls.awakeable(typeTag, cb));

    return new Awakeable<>(syscalls, awakeable.getValue(), awakeable.getKey());
  }

  @Override
  public <T> void completeAwakeable(
      AwakeableIdentifier id, TypeTag<T> typeTag, @Nonnull T payload) {
    Util.<Void>blockOnSyscall(cb -> syscalls.completeAwakeable(id, typeTag, payload, cb));
  }
}

File: sdk-kotlin/src/main/kotlin/dev/restate/sdk/kotlin/RestateContextImpl.kt
package dev.restate.sdk.kotlin

import com.google.protobuf.MessageLite
import dev.restate.sdk.core.*
import dev.restate.sdk.core.syscalls.DeferredResult
import dev.restate.sdk.core.syscalls.EnterSideEffectSyscallCallback
import dev.restate.sdk.core.syscalls.ExitSideEffectSyscallCallback
import dev.restate.sdk.core.syscalls.Syscalls
import io.grpc.MethodDescriptor
import io.grpc.StatusRuntimeException
import kotlin.coroutines.resume
import kotlin.time.Duration
import kotlin.time.toJavaDuration
import kotlinx.coroutines.*

internal class RestateContextImpl internal constructor(private val syscalls: Syscalls) :
    RestateContext {
  override suspend fun <T : Any> get(key: StateKey<T>): T? {
    val deferredResult: DeferredResult<T> =
        suspendCancellableCoroutine { cont: CancellableContinuation<DeferredResult<T>> ->
          syscalls.get(key.name(), key.typeTag(), completingContinuation(cont))
        }

    if (!deferredResult.isCompleted) {
      suspendCancellableCoroutine { cont: CancellableContinuation<Unit> ->
        syscalls.resolveDeferred(deferredResult, completingUnitContinuation(cont))
      }
    }

    val readyResult = deferredResult.toReadyResult()!!
    if (!readyResult.isSuccess) {
      throw readyResult.failure!!
    }
    if (readyResult.isEmpty) {
      return null
    }
    return readyResult.result!!
  }

  override suspend fun <T : Any> set(key: StateKey<T>, value: T) {
    return suspendCancellableCoroutine { cont: CancellableContinuation<Unit> ->
      syscalls.set(key.name(), key.typeTag(), value, completingUnitContinuation(cont))
    }
  }

  override suspend fun clear(key: StateKey<*>) {
    return suspendCancellableCoroutine { cont: CancellableContinuation<Unit> ->
      syscalls.clear(key.name(), completingUnitContinuation(cont))
    }
  }

  override suspend fun timer(duration: Duration): Awaitable<Unit> {
    val deferredResult: DeferredResult<Void> =
        suspendCancellableCoroutine { cont: CancellableContinuation<DeferredResult<Void>> ->
          syscalls.sleep(duration.toJavaDuration(), completingContinuation(cont))
        }

    return UnitAwaitableImpl(syscalls, deferredResult)
  }

  override suspend fun <T : MessageLite, R : MessageLite> callAsync(
      methodDescriptor: MethodDescriptor<T, R>,
      parameter: T
  ): Awaitable<R> {
    val deferredResult: DeferredResult<R> =
        suspendCancellableCoroutine { cont: CancellableContinuation<DeferredResult<R>> ->
          syscalls.call(methodDescriptor, parameter, completingContinuation(cont))
        }

    return NonNullAwaitableImpl(syscalls, deferredResult)
  }

  override suspend fun <T : MessageLite, R : MessageLite> oneWayCall(
      methodDescriptor: MethodDescriptor<T, R>,
      parameter: T
  ) {
    return suspendCancellableCoroutine { cont: CancellableContinuation<Unit> ->
      syscalls.backgroundCall(methodDescriptor, parameter, null, completingUnitContinuation(cont))
    }
  }

  override suspend fun <T : MessageLite, R : MessageLite> delayedCall(
      methodDescriptor: MethodDescriptor<T, R>,
      parameter: T,
      delay: Duration
  ) {
    return suspendCancellableCoroutine { cont: CancellableContinuation<Unit> ->
      syscalls.backgroundCall(
          methodDescriptor, parameter, delay.toJavaDuration(), completingUnitContinuation(cont))
    }
  }

  override suspend fun <T : Any?> sideEffect(
      typeTag: TypeTag<T>,
      sideEffectAction: suspend () -> T
  ): T {
    val exitResult =
        suspendCancellableCoroutine { cont: CancellableContinuation<CompletableDeferred<T>> ->
          syscalls.enterSideEffectBlock(
              typeTag,
              object : EnterSideEffectSyscallCallback<T> {
                @Suppress("UNCHECKED_CAST")
                override fun onResult(t: T?) {
                  val deferred: CompletableDeferred<T> = CompletableDeferred()
                  // This unchecked cast is fine because T is declared as Any?
                  deferred.complete(t as T)
                  cont.resume(deferred)
                }

                override fun onFailure(t: StatusRuntimeException) {
                  val deferred: CompletableDeferred<T> = CompletableDeferred()
                  deferred.completeExceptionally(t)
                  cont.resume(deferred)
                }

                override fun onCancel(t: Throwable?) {
                  cont.cancel(t)
                }

                override fun onNotExecuted() {
                  cont.resume(CompletableDeferred())
                }
              })
        }

    if (exitResult.isCompleted) {
      return exitResult.await()
    }

    var actionReturnValue: T? = null
    var actionFailure: Throwable? = null
    try {
      actionReturnValue = sideEffectAction()
    } catch (e: Throwable) {
      actionFailure = e
    }

    val exitCallback =
        object : ExitSideEffectSyscallCallback<T> {
          @Suppress("UNCHECKED_CAST")
          override fun onResult(t: T?) {
            // This unchecked cast is fine because T is declared as Any?
            exitResult.complete(t as T)
          }

          override fun onFailure(t: StatusRuntimeException) {
            exitResult.completeExceptionally(t)
          }

          override fun onCancel(t: Throwable?) {
            exitResult.cancel(CancellationException("Suspended", t))
          }
        }

    if (actionFailure != null) {
      syscalls.exitSideEffectBlockWithException(actionFailure, exitCallback)
    } else {
      syscalls.exitSideEffectBlock(typeTag, actionReturnValue, exitCallback)
    }

    return exitResult.await()
  }

  override suspend fun <T> awakeable(typeTag: TypeTag<T>): Awakeable<T> {
    val (aid, deferredResult) =
        suspendCancellableCoroutine {
            cont: CancellableContinuation<Map.Entry<String, DeferredResult<T>>> ->
          syscalls.awakeable(typeTag, completingContinuation(cont))
        }

    return AwakeableImpl(syscalls, deferredResult, aid)
  }

  override fun awakeableHandle(id: String): AwakeableHandle {
    return AwakeableHandleImpl(syscalls, id)
  }

  override suspend fun compensate(compensation: suspend () -> Unit) {
    val scope = CoroutineScope(currentCoroutineContext())
    val compensationJob = scope.async(start = CoroutineStart.LAZY) { compensation() }

    syscalls.registerCompensation { syscallCallback ->
      compensationJob.start()
      compensationJob.invokeOnCompletion {
        if (it == null) {
          syscallCallback.onSuccess(null)
        } else {
          syscallCallback.onCancel(it)
        }
      }
    }
  }
}
